static/js/stream_list.js

```diff
@@ -420,6 +420,15 @@ export function update_streams_sidebar(force_rerender) {
 }
 
 export function update_dom_with_unread_counts(counts) {
+    if (!counts.stream_count.size) {
+        // No messages are unread in any stream.
+        // To ensure the UI reflects the same, update all subscribed streams
+        // with 0 unread count.
+        const streams = stream_data.subscribed_stream_ids();
+        for (const stream of streams) {
+            set_stream_unread_count(stream, 0, false);
+        }
+    }
```

Comment:
This feels too much like a special case. The purpose of this function is to set the unread count for each stream correctly; the bug here is that it doesn't set to 0 . It sounds like `counts.stream_count` only contains streams with nonzero unreads; so I think we should collect the set of streams that we've visited through that, and then after the existing `counts.stream_count` loop, do a loop through `stream_data.subscribed_stream_ids` and call `set_stream_unread_count(stream, 0, false);` for those...

Reply:
I guess I'm also not sure I understand why the existing logic works at all. Maybe the situation is that `counts.stream_count` normally contains every stream that has __ever__ had an unread in this browser session, since we don't delete buckets when they get down to 0 items in them; but the `declare_bankruptcy` function violates that invariant?

Reply:
> but the declare_bankruptcy function violates that invariant, by removing the buckets in the bucketer?

Yes yes.